### PR TITLE
Workflow to check for labels

### LINKED ISSUE #742: Add a gh actions labelling check workflow
Followup to https://github.com/Manta-Network/Manta/issues/661

### PR DESCRIPTION
**Pull request by Apokalip**
Signed-off-by: Apokalip <simeon@manta.network>

## Workflow/s to check for labels in branches and release branches
closes: #742 
relates: #661

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Added **one** label out of the `L-` group to this PR
- [x] Added **one or more** labels out of `A-` and `C-` groups to this PR
- [x] Re-reviewed `Files changed` in the Github PR explorer.

Situational Notes:
- If adding functionality, write unit tests!
- If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.
- If runtime changes, need to update the version numbers properly:
   * `authoring_version`: The version of the authorship interface. An authoring node will not attempt to author blocks unless this is equal to its native runtime.
   * `spec_version`: The version of the runtime specification. A full node will not attempt to use its native runtime in substitute for the on-chain Wasm runtime unless all of spec_name, spec_version, and authoring_version are the same between Wasm and native.
   * `impl_version`: The version of the implementation of the specification. Nodes are free to ignore this; it serves only as an indication that the code is different; as long as the other two versions are the same then while the actual code may be different, it is nonetheless required to do the same thing. Non-consensus-breaking optimizations are about the only changes that could be made which would result in only the impl_version changing.
   * `transaction_version`: The version of the extrinsics interface. This number must be updated in the following circumstances: extrinsic parameters (number, order, or types) have been changed; extrinsics or pallets have been removed; or the pallet order in the construct_runtime! macro or extrinsic order in a pallet has been changed. You can run the `metadata_diff.yml` workflow for help. If this number is updated, then the `spec_version` must also be updated
- Verify benchmarks & weights have been updated for any modified runtime logics

### CODE REVIEW

File: .github/workflows/label_checker.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  pull_request:
+    branches: [manta]
+jobs:
+  check_l_labels:
+    name: L- Labels Check
+    runs-on: ubuntu-20.04
+    steps:
+      - uses: docker://agilepathway/pull-request-label-checker:latest
+        with:
+          any_of: |
```
Comment (Garandor) [inline .github/workflows/label_checker.yml]:
```suggestion
          one_of: |
```

**Review by Garandor — DISMISSED** [review]
Thanks for the quick fix!

Does this action support `*` wildcards?
If so you could make it `one_of: L-*` and add C- and A- checks with `any_of: A-*`

**Comment by Apokalip** [comment]
@Garandor Unfortunately I could not find a label action that does this with wild cards or regex. If I had the time I would write it on my own or fork this one.

File: .github/workflows/release_checks.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    tags:
+      - 'v[0-9]+\.[0-9]+\.[0-9]+(\-[0-9]+)?'
+jobs:
+  release_label:
+    name: Release Label Check
+    runs-on: ubuntu-latest
+    steps:
+      - uses: docker://agilepathway/pull-request-label-checker:latest
+        with:
+          all_of: L-release
```
Comment (Garandor) [inline .github/workflows/release_checks.yml]: are you sure we want a release label? i though a release PR will just get `L-skip` to not show in the Changelog.

Also - i just realize that a `L-skip` label doesn't seem to exist ⭕

Comment (Apokalip) [reply .github/workflows/release_checks.yml]: Why do we need a L-skip when if you dont have a Label, the PR will be skipped
And I got the L-release from #661 the TODO

Comment (Garandor) [reply .github/workflows/release_checks.yml]: if you check for `one-of: L-` the check will fail if no `L-` tag is present. By design, to remind people to add the label.

So if you intentionally want no changelog entry from a PR, a `L-skip` is needed

Comment (Apokalip) [reply .github/workflows/release_checks.yml]: I will add it back

**Comment by Garandor** [comment]
I still don't get the purpose of the `L-release` label.
A release branch is already identified by its `release-` branch name, why does it need a label?

**Comment by Apokalip** [comment]
> I still don't get the purpose of the `L-release` label. A release branch is already identified by its `release-` branch name, why does it need a label?

 In the original issue #661 the TODO part, I am 50/50 about it honestly

**Comment by Garandor** [comment]
> In the original issue #661 the TODO part, I am 50/50 about it honestly

Oh that's where you got the requirement from. I actually don't remember the specific reason for this design constraint myself.
I *think* knowing a `release-xyz` is a release branch is good enough.
Also, not having a `L-release` will allow to enforce the `one of L-` rule, otherwise a release could have `L-release` and `L-skip` violating that rule ( anbd the generator tool would need to be written to ignore both `release` and `skip` labels in the changelog).

**Comment by Apokalip** [comment]
> > In the original issue #661 the TODO part, I am 50/50 about it honestly
> 
> Oh that's where you got the requirement from. I actually don't remember the specific reason for this design constraint myself. I _think_ knowing a `release-xyz` is a release branch is good enough. Also, not having a `L-release` will allow to enforce the `one of L-` rule, otherwise a release could have `L-release` and `L-skip` violating that rule ( anbd the generator tool would need to be written to ignore both `release` and `skip` labels in the changelog).

the generator will skip releases with extension as v3.2.0-1, unless you change the default config regex ofc

**Comment by Garandor** [comment]
I think the point is, the release branch will be merged into manta eventually, becoming a commit on the `manta` branch
If then we git tag that commit, ~~it'll run the verifier again~~ and notice there's a new commit - the release itself - which is not part of the changelog, no matter what the version number on the release branch used to be.

~~Won't that fail?~~
Actually nevermind, the chaneglog verifier doesn't run on commits to manta branch, so we're safe there

**Comment by Garandor** [comment]
Ok, so while I still see a need for `L-skip` in order to support minor PRs like formatting changes, i see no reason for `L-release` to exist.

Looking for other opinions on this @bhgomes @Dengjianping @ghzlatarev

**Comment by ghzlatarev** [comment]
> Ok, so while I still see a need for `L-skip` in order to support minor PRs like formatting changes, i see no reason for `L-release` to exist.
> 
> Looking for other opinions on this @bhgomes @Dengjianping @ghzlatarev

remove l-release, that's fine.

**Review by ghzlatarev — DISMISSED** [review]

File: .github/workflows/label_checker.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    runs-on: ubuntu-20.04
+    steps:
+      - uses: docker://agilepathway/pull-request-label-checker:latest
+        with:
+          one_of: |
+            L-added,
+            L-changed,
+            L-fixed,
+            L-deprecated,
+            L-removed
```
Comment (Garandor) [inline .github/workflows/label_checker.yml]: add `L-skip` and we're good to go

**Review by Garandor — DISMISSED** [review]

**Comment by Garandor** [comment]
image

this seems weird, why does this run in 4 different workflows?

**Comment by Apokalip** [comment]
> image
> 
> this seems weird, why does this run in 4 different workflows?

those are different checks, see the underlying text under the title, linter, tests, publish draft etc

**Comment by Garandor** [comment]
> > image
> > this seems weird, why does this run in 4 different workflows?
> 
> those are different checks, see the underlying text under the title, linter, tests, publish draft etc

my bad, was reading this wrong

**Review by Garandor — APPROVED** [review]
LGTM!

**Review by ghzlatarev — APPROVED** [review]

**Comment by Garandor** [comment]
Oh, we can also use this PR to automatically prevent PRs that are labelled as `DO-NOT-MERGE` from being merged.

Another change i'd request is to make passing this workflow `Required` for all PRs

**Review by stechu — APPROVED** [review]

**Pull request merged**